File: LC/[0121] Best Time to Buy and Sell Stock.cpp
class Solution {
public:

    int maxProfit(vector<int>& prices) {
        int min_so_far=prices[0];
        int max_p_so_far=0;

        for(int ii=0; ii<prices.size(); ++ii){
            min_so_far=min(min_so_far, prices[ii]);
            max_p_so_far=max(max_p_so_far, prices[ii]-min_so_far);
        }

        return max_p_so_far;
    }
};

File: LC/[0121] Best Time to Buy and Sell Stock.py

#####################
# 20230114
#####################
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        : carry a min sofar value
        : profit=max(p[ii]-minSoFar, profit)
        : minSoFar=min(minSoFar, p[ii])
        """
        profit=0
        minSoFar=prices[0]
        for p in prices:
            profit=max(profit, p-minSoFar)
            minSoFar=min(p, minSoFar)

        return max(0, profit)


#####################
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_so_far=prices[0]
        mx_profit=0
        for p in prices:
            mx_profit=max(mx_profit, p-min_so_far)
            min_so_far=min(min_so_far, p)

        return mx_profit
